Alerting: Rule API

yuri-tceretian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
This PR is a preparation for implementing support for fine-grained access control in rule API. The current implementation of the rule API allow users to perform many opertaions in one request (create, update or delete a rule). The API model is processed by the data layer and therefore it is not possible to perform authorization in the controller.
In this PR I try to separate responsibilities between controller and data layer:
- Add validation logic for API models. It copies what's done in storage layer with some additional validations. The result of validation is either user-friendly error or a collection of AlertRule. This allows us to validate model before opening the transaction.
- Update data store's method UpsertRules to not populate the new rule from the existing rule. This is not necessary because API layer expects complete model of the rule. Otherwise it will fail validation.
- Added function that pre-calculates changes before updating the database. The result of the function is a collection of rules that needs to be added, updated and removed. Using these calculations I will add authorization logic in future PR. 
- Moved transaction out of storage layer. Changes made in https://github.com/grafana/grafana/pull/44939 allowed reusing session and abstract from Sql client. 
- The refactoring made possible to perform more correct quota validation. It is moved to the end of the transaction to consider state after all changes. 

**Special notes for your reviewer**:

yuri-tceretian committed:
add validation of rules API model

yuri-tceretian committed:
add function to calculate changes between the submitted alerts and existing alerts

yuri-tceretian committed:
update RoutePostNameRulesConfig to validate input models, calculate changes and apply in a transaction

yuri-tceretian committed:
delete unused storage method

gotjosh commented:
@papagian can you take a look first? You have more context on this than I do.

papagian commented on pkg/services/ngalert/api/api_ruler.go:
```diff
@@ -243,62 +249,73 @@ func (srv RulerSrv) RoutePostNameRulesConfig(c *models.ReqContext, ruleGroupConf
-		}
-		if limitReached {
-			return ErrResp(http.StatusForbidden, errors.New("quota reached"), "")
+		if len(changes.Upsert) > 0 {
```
any reason that you don't check `changes.newRules > 0` instead? if all rules in `changes.Upsert` are updates (no new rules) then I think there is no need to check the quota.

yuri-tceretian replied:
fixed.

papagian commented on pkg/services/ngalert/api/api_ruler.go:
```diff
@@ -243,62 +249,73 @@ func (srv RulerSrv) RoutePostNameRulesConfig(c *models.ReqContext, ruleGroupConf
-				return ErrResp(http.StatusBadRequest, fmt.Errorf("conflicting UID %q found", r.GrafanaManagedAlert.UID), "failed to validate alert rule %q", r.GrafanaManagedAlert.Title)
+
+		for _, rule := range changes.Delete {
+			if err = srv.store.DeleteAlertRuleByUID(tranCtx, c.SignedInUser.OrgId, rule.UID); err != nil {
```
nit: we could delete all the rules in one SQL query

yuri-tceretian replied:
Makes sense! I will address that in the future PR

papagian commented on pkg/services/ngalert/ngalert.go:
```diff
@@ -94,9 +94,9 @@ func (ng *AlertNG) init() error {
 	if baseInterval <= 0 {
-		baseInterval = defaultBaseIntervalSeconds
+		baseInterval = defaultBaseIntervalSeconds * time.Second
+		ng.Cfg.AlertingBaseInterval = baseInterval
```
nit, `AlertingBaseInterval` should better move under `Cfg.UnifiedAlerting`.
I have noticed also that we never `ng.Cfg.AlertingBaseInterval` so it always gets the default one.

yuri-tceretian replied:
Yes. this is a hack until https://github.com/grafana/grafana/pull/45107 is  merged. I would also appreciate your review of it.

papagian reviewed (approved):
nice! I wonder whether we still need `DBStore.validateAlertRule()`

yuri-tceretian committed:
fix AlertForDuration to DefaultRuleEvaluationInterval

because it correctly reflects the purpose of the field

yuri-tceretian committed:
add InTransaction to RuleStore

yuri-tceretian committed:
add config to RuleSvr

yuri-tceretian committed:
add tests for alert rule model

yuri-tceretian committed:
add tests for calculate changes

yuri-tceretian committed:
fix creation of test alert rule

yuri-tceretian committed:
fix messages in integration test

yuri-tceretian committed:
fail if UID is specified but no rule is found in db

this resembles the way it works in Upsert

yuri-tceretian committed:
simplify upsert

- update upsert to not modify fields of new by values from the existing alert
- if rule has UID do not try to pull it from db. (it is done upstream)

yuri-tceretian committed:
updated fakeRuler's PutRule to accept many rules

yuri-tceretian committed:
calculate changes should not modify rules

yuri-tceretian committed:
add rule generator

yuri-tceretian committed:
more tests for calculateChanges

yuri-tceretian committed:
fix messages and return proper status

yuri-tceretian committed:
return support for patching of models

yuri-tceretian committed:
fixup

yuri-tceretian committed:
better message for unique index violation

yuri-tceretian committed:
damn linter

yuri-tceretian committed:
Merge branch 'main' into rule-mgmt-in-ctrl

# Conflicts:
#	pkg/setting/setting_unified_alerting.go

JohnnyQQQQ reviewed (approved):
👍
